### PR TITLE
Source.actorRef with buffer stats (#25402)

### PR DESCRIPTION
**Pull request by tg44**
Hy!

A small PR with a small QOL improvement. It can be useful for the early stages of development, under loadtests, and can be used in production as metrics.

fixes #25402

### CODE REVIEW

File: akka-docs/src/test/java/jdocs/stream/operators/ActorRefSource.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import akka.stream.*;
+import static akka.pattern.PatternsCS.ask;
+
+import akka.stream.javadsl.Sink;
+import akka.stream.javadsl.Source;
+import akka.util.Timeout;
+
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+//#sourceActorRefBufferSize
```
Comment (patriknw) [inline akka-docs/src/test/java/jdocs/stream/operators/ActorRefSource.java]: additional empty line, to separate the imports from the rest of the code

File: akka-docs/src/test/scala/docs/stream/operators/SourceOperators.scala [hunk 2]
```diff
@@ -25,4 +25,27 @@ object SourceOperators {
...
+    import concurrent.duration._
+    import akka.util.Timeout
+
+    implicit val system: ActorSystem = ActorSystem()
+    implicit val materializer: ActorMaterializer = ActorMaterializer()
+    implicit val executionContext = system.dispatcher
+    implicit val timeout = Timeout(5.seconds)
+
+    val ref: ActorRef = Source.actorRef(1000, OverflowStrategy.fail).to(Sink.ignore).run
+    val ans: Future[BufferStatus] = (ref ? GetStatusMessage()).mapTo[BufferStatus]
```
Comment (patriknw) [inline akka-docs/src/test/scala/docs/stream/operators/SourceOperators.scala]: rename to `GetBufferStatus`?

File: akka-stream/src/main/scala/akka/stream/BufferStatus.scala [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+sealed abstract class GetStatusMessage
+
+object GetStatusMessage {
+  val instance = new GetStatusMessage {}
+
+  def apply(): GetStatusMessage = instance
+
+  def getStatusMessage(): GetStatusMessage = instance
+}
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/BufferStatus.scala]: we typically define such singleton objects as:

```scala
@DoNotInherit sealed abstract class GetStatusMessage

case object GetStatusMessage extends GetStatusMessage {
  /**
   * Java API: the singleton instance
   */
  def getInstance(): GetStatusMessage = this
}
```

and when matching `case GetStatusMessage =>`

File: akka-stream/src/main/scala/akka/stream/impl/ActorRefSourceActor.scala [hunk 4]
```diff
@@ -117,4 +117,9 @@ import akka.stream.ActorMaterializerSettings
         "only draining already buffered elements: [{}] (pending: [{}])", elem, buffer.used)
   }
 
+  def bufferStatusRequest: Receive = {
+    case _: GetStatusMessage ⇒
+      sender ! new BufferStatus(buffer.used, buffer.capacity)
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/ActorRefSourceActor.scala]: `sender()`

Comment (patriknw) [reply akka-stream/src/main/scala/akka/stream/impl/ActorRefSourceActor.scala]: note that `buffer` can be `null`, see comment where it is declared

**Comment by tg44** [comment]
Thanks the review. Fixed!

**Comment by johanandren** [comment]
Hi @tg44 I'm afraid the `ActorRefSource` has recently been rewritten as a graphstage and while we appreciate the work you put into this it will not be possible to merge so I will close it for now.

**Pull request closed without merging**